sql-plugins/src/main/java/io/cdap/pipeline/sql/plugins/joiner/Joiner.java

```diff
@@ -0,0 +1,277 @@
+      conditions.add(condition);
+    }
+    return builder.join(calciteJoinType, conditions);
+  }
+
+  @Override
+  public RelNode getQuery(RelBuilder builder) {
+    List<RexNode> fields = new ArrayList<>();
+    for (Map.Entry<String, String> selectedFieldsEntry: selectedFields.entrySet()) {
+      String fieldName = selectedFieldsEntry.getKey();
+      String stageName = fieldName.split(Pattern.quote("."))[0];
+      if (!stageOrdinalMap.containsKey(stageName)) {
+        throw new IllegalArgumentException("Selected field stage is not in the list of input stages.");
+      }
+      String alias = selectedFieldsEntry.getValue();
+      RexNode field = builder.alias(builder.field(fieldName), alias);
+      fields.add(field);
+    }
+    return builder.project(fields).build();
+  }
+
+  private void init(List<String> inputs) {
+    selectedFields = new LinkedHashMap<>();
+    joinKeys = new ArrayList<>();
+    // Get a map of the ordinals
+    stageOrdinalMap = new HashMap<>();
+    for (int i = 0; i < inputs.size(); i++) {
+      stageOrdinalMap.put(inputs.get(i), i);
+    }
+
+    // Parse the config strings
+    parseSelectedFields();
+    parseJoinKeys();
+
+    // Parse the join type
+    try {
+      calciteJoinType = JoinRelType.valueOf(config.getJoinType().toUpperCase());
+    } catch (IllegalArgumentException e) {
+      throw new IllegalArgumentException("Unsupported join type: " + config.getJoinType());
+    }
+  }
+
+  /**
+   * Parses the selected fields into a list.
+   */
+  private void parseSelectedFields() {
+    if (Strings.isNullOrEmpty(config.getSelectedFields())) {
+      throw new IllegalArgumentException("Selected fields cannot be empty.");
+    }
+    String[] fields = config.getSelectedFields().trim().split("\\s*,\\s*");
+    for (String field: fields) {
+      String[] fieldAliasArr = field.split(Pattern.quote(" as "));
+      if (fieldAliasArr.length != 2 || Strings.isNullOrEmpty(fieldAliasArr[0])
+        || Strings.isNullOrEmpty(fieldAliasArr[1])) {
+        throw new IllegalArgumentException("Selected fields must contain a field name and an alias.");
+      }
+      String fieldName = fieldAliasArr[0];
+      String alias = fieldAliasArr[1];
+      String[] fieldArr = fieldAliasArr[0].split(Pattern.quote("."));
+      if (fieldArr.length != 2 || Strings.isNullOrEmpty(fieldArr[0]) || Strings.isNullOrEmpty(fieldArr[1])) {
+        throw new IllegalArgumentException("Selected fields must contain a stage name and a field name.");
+      }
+      if (selectedFields.containsKey(fieldName)) {
+        throw new IllegalArgumentException("Unable to rename a field twice.");
+      }
+      if (selectedFields.containsValue(alias)) {
+        throw new IllegalArgumentException("Unable to rename two fields to the same name.");
+      }
+      selectedFields.put(fieldName, alias);
+    }
+  }
+
+  /**
+   * Parses the join keys string from the config into {@link JoinKey} objects.
+   */
+  private void parseJoinKeys() {
```

Comment:
this should go in the config class and return the list of JoinKeys instead of modifying state.

Reply:
Fixed